config/config.go

```diff
@@ -1,41 +1,248 @@
+
+func SetupConfig() {
+	if !verifyConfigCmd() || !isConfigFilePresent() {
+		DiceConfig = &defaultConfig
+		log.Info("Using default configurations.")
+		return
+	}
+
+	setUpViperConfig(ConfigFileLocation)
+}
+
+func verifyConfigCmd() bool {
+	if areBothFlagsSet() {
+		log.Error("Both -o and -c flags are set. Please use only one flag.")
+		return false
+	}
+
+	// if -o flag is set, create a config file at the specified path by verifying the path
+	if isValidDirPath() {
+		ConfigFileLocation = filepath.Join(CustomConfigFilePath, DefaultConfigName)
+		createConfigFile(ConfigFileLocation)
+		return true
+	}
+
+	// if -c flag is set then viper will take care of reading the config file and verifying it
+	return true
+}
+
+func createConfigFile(configFilePath string) {
+	if _, err := os.Stat(configFilePath); err == nil {
+		log.Warnf("config file already exists at %s", configFilePath)
+		setUpViperConfig(configFilePath)
+		return
+	}
+
+	if err := writeConfigFile(configFilePath); err != nil {
+		log.Error(err)
+		log.Warn("starting DiceDB with default configurations.")
+		DiceConfig = &defaultConfig
+		return
+	}
+
+	log.Infof("config file created at %s with default configurations", configFilePath)
+}
+
+func writeConfigFile(configFilePath string) error {
+	dir := filepath.Dir(configFilePath)
+	if _, err := os.Stat(dir); err != nil {
+		return err
+	}
+
+	log.Infof("creating default config file at %s", configFilePath)
+	file, err := os.Create(configFilePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	encoder := toml.NewEncoder(file)
+	err = encoder.Encode(defaultConfig)
+	return err
+}
+
+func isValidDirPath() bool {
+	if CustomConfigFilePath == utils.EmptyStr {
+		return false
+	}
+
+	info, err := os.Stat(CustomConfigFilePath)
+	if os.IsNotExist(err) || err != nil {
+		return false
+	}
+
+	if !info.IsDir() {
+		return false
+	}
+	return true
+}
+
+func areBothFlagsSet() bool {
+	return CustomConfigFilePath != utils.EmptyStr && ConfigFileLocation != filepath.Join(DefaultConfigFilePath, DefaultConfigName)
+}
+
+func setUpViperConfig(configFilePath string) {
+	viper.SetConfigName("dice")
+
+	if configFilePath == utils.EmptyStr {
+		viper.AddConfigPath(DefaultConfigFilePath)
+	} else {
+		viper.AddConfigPath(filepath.Dir(configFilePath))
+	}
+
+	viper.SetConfigType("toml")
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			log.Warn("config file not found. Using default configurations.")
+			DiceConfig = &defaultConfig
+			return
+		}
+		log.Errorf("Error reading config file: %v", err.Error())
+	}
+
+	// this line ensures that if config file is partially filled then the rest of the fields will be filled with default values
+	DiceConfig = &defaultConfig
+
+	if err := viper.Unmarshal(&DiceConfig); err != nil {
+		log.Errorf("Error unmarshalling config file: %v", err.Error())
+		log.Warn("starting DiceDB with default configurations.")
+		DiceConfig = &defaultConfig
```

Comment:
Rather than multiple default assignments, shall we assign defaults intially and override them as the setup via viper is done?

Reply:
Indeed